src/run_graph.rs

```diff
@@ -79,66 +78,106 @@ impl TasksGraph {
         id
     }
 
-    pub fn next_task(&mut self) -> Option<(NodeIndex, Task)> {
+    pub fn next_task<DB: WriteResults>(
+        &mut self,
+        ex: &Experiment,
+        db: &DB,
+    ) -> Option<(NodeIndex, Arc<Task>)> {
         let root = self.root;
-        if let WalkResult::Task(id, task) = self.walk_graph(root) {
+        if let WalkResult::Task(id, task) = self.walk_graph(root, ex, db) {
             Some((id, task))
         } else {
             None
         }
     }
 
-    fn walk_graph(&mut self, node: NodeIndex) -> WalkResult {
+    fn walk_graph<DB: WriteResults>(
+        &mut self,
+        node: NodeIndex,
+        ex: &Experiment,
+        db: &DB,
+    ) -> WalkResult {
+        // Ensure tasks are only executed if needed
+        if let Node::Task {
+            ref task,
+            running: false,
+        } = self.graph[node]
+        {
+            if !task.needs_exec(ex, db) {
+                return WalkResult::NotBlocked;
+            }
```

Comment:
So `!needs_exec` will only return true if a crate result is already completed according to the database, but for some reason the graph doesn't reflect it (perhaps when resuming an experiment?). In this case, shouldn't we take the opportunity to remove the node from the graph so we don't keep traversing over it?